modules/core/multihop/proof_gen.go

```diff
@@ -35,26 +37,58 @@ type Endpoint interface {
 	Counterparty() Endpoint
 }
 
+func EndpointToString(endpoint Endpoint) string {
+	chainHeight := endpoint.GetChainHeight()
+	kvHeight := endpoint.GetKeyValueProofHeight()
+	consHeight := endpoint.GetConsensusHeight()
+	return fmt.Sprintf("{\"chain_id\": %q, \"client_id\": %q, \"connection_id\": %q, \"chain_height\": %d, \"kv_proof_height\": %q, \"consensus_height\": %q}",
+		endpoint.ChainID(), endpoint.ClientID(), endpoint.ConnectionID(), chainHeight, kvHeight, consHeight)
+}
+
 // Path contains two endpoints of chains that have a direct IBC connection, ie. a single-hop IBC path.
 type Path struct {
 	EndpointA Endpoint
 	EndpointB Endpoint
 }
 
+func (p Path) String() string {
+	return fmt.Sprintf("[%s, %s]", EndpointToString(p.EndpointA), EndpointToString(p.EndpointB))
+}
+
 // ChanPath represents a multihop channel path that spans 2 or more single-hop `Path`s.
-type ChanPath []*Path
+type ChanPath struct {
+	Paths        []*Path
+	counterparty *ChanPath
```

Comment:
Can we keep the old ChanPath definition since counterparty is not used in proof gen?